File: core/src/query/parse.rs
//! The Rust parser, the second consumer of §8.3's one production.

use super::ast::{
    Cmp, HasAttribute, IgnoredReason, IgnoredTerm, IsFlag, QueryAst, QueryTerm, TextField,
};
use super::QUERY_GRAMMAR_VERSION;

/// §8.3: quoting is `"…"` with `\"` as the only escape. Whitespace inside quotes does not split.
fn tokenize(input: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;
    let mut escaped = false;
    for ch in input.chars() {
        if escaped {
            current.push(ch);
            escaped = false;
            continue;
        }
        match ch {
            '\\' if in_quotes => {
                current.push(ch);
                escaped = true;
            }
            '"' => {
                in_quotes = !in_quotes;
                current.push(ch);
            }
            c if c.is_whitespace() && !in_quotes => {
                if !current.is_empty() {
                    tokens.push(std::mem::take(&mut current));
                }
            }
            c => current.push(c),
        }
    }
    if !current.is_empty() {
        tokens.push(current);
    }
    tokens
}

fn unquote(value: &str) -> (String, bool) {
    let mut chars = value.chars();
    if chars.next() == Some('"') {
        if value.chars().count() >= 2 && value.ends_with('"') {
            let inner: String = chars.clone().take(value.chars().count() - 2).collect();
            return (inner.replace("\\\"", "\""), true);
        }
        // Unterminated: take what is there rather than dropping the term silently.
        return (chars.collect::<String>().replace("\\\"", "\""), true);
    }
    (value.to_owned(), false)
}

/// A prefix that reads as a field name: two or more letters. One letter is a drive letter.
fn is_field_shaped(prefix: &str) -> bool {
    prefix.chars().count() >= 2 && prefix.chars().all(|c| c.is_ascii_alphabetic())
}

fn parse_is(value: &str) -> Option<IsFlag> {
    Some(match value {
        "dirty" => IsFlag::Dirty,
        "unpushed" => IsFlag::Unpushed,
        "behind" => IsFlag::Behind,
        "interrupted" => IsFlag::Interrupted,
        "archived" => IsFlag::Archived,
        "pinned" => IsFlag::Pinned,
        "hidden" => IsFlag::Hidden,
        "reference" => IsFlag::Reference,
        "bare" => IsFlag::Bare,
        "fork" => IsFlag::Fork,
        "empty" => IsFlag::Empty,
        "shallow" => IsFlag::Shallow,
        "local" => IsFlag::Local,
        "wsl" => IsFlag::Wsl,
        "new" => IsFlag::New,
        _ => return None,
    })
}

fn parse_has(value: &str) -> Option<HasAttribute> {
    Some(match value {
        "readme" => HasAttribute::Readme,
        "license" => HasAttribute::License,
        "tests" => HasAttribute::Tests,
        "ci" => HasAttribute::Ci,
        "remote" => HasAttribute::Remote,
        "stash" => HasAttribute::Stash,
        "submodules" => HasAttribute::Submodules,
        _ => return None,
    })
}

/// `>10mb` / `<6mo` becomes (op, digits, unit). Nothing else is a comparison.
fn split_comparison(value: &str) -> Option<(Cmp, u64, String)> {
    let mut chars = value.chars();
    let op = match chars.next()? {
        '>' => Cmp::Gt,
        '<' => Cmp::Lt,
        _ => return None,
    };
    let rest: String = chars.collect();
    let digits: String = rest.chars().take_while(char::is_ascii_digit).collect();
    let unit: String = rest.chars().skip(digits.chars().count()).collect();
    if digits.is_empty() || unit.is_empty() || !unit.chars().all(|c| c.is_ascii_lowercase()) {
        return None;
    }
    Some((op, digits.parse().ok()?, unit))
}

fn parse_field_term(field: &str, raw: &str, quoted: bool, negated: bool) -> Option<QueryTerm> {
    if raw.is_empty() {
        return None;
    }
    match field {
        "lang" | "owner" | "in" | "collection" => {
            let value = if quoted {
                raw.to_owned()
            } else {
                raw.to_lowercase()
            };
            let text_field = match field {
                "lang" => TextField::Lang,
                "owner" => TextField::Owner,
                "in" => TextField::In,
                _ => TextField::Collection,
            };
            Some(QueryTerm::Text {
                negated,
                field: text_field,
                value,
                quoted,
            })
        }
        "is" => parse_is(&raw.to_lowercase()).map(|flag| QueryTerm::Flag { negated, flag }),
        "has" => {
            parse_has(&raw.to_lowercase()).map(|attribute| QueryTerm::Has { negated, attribute })
        }
        "size" => {
            let (op, n, unit) = split_comparison(&raw.to_lowercase())?;
            let scale: u64 = match unit.as_str() {
                "kb" => 1024,
                "mb" => 1_048_576,
                "gb" => 1_073_741_824,
                _ => return None,
            };
            Some(QueryTerm::Size {
                negated,
                op,
                bytes: n.checked_mul(scale)?,
            })
        }
        "touched" => {
            let lowered = raw.to_lowercase();
            if lowered.chars().count() == 4 && lowered.chars().all(|c| c.is_ascii_digit()) {
                return lowered
                    .parse()
                    .ok()
                    .map(|year| QueryTerm::TouchedYear { negated, year });
            }
            let (op, n, unit) = split_comparison(&lowered)?;
            let scale: u64 = match unit.as_str() {
                "d" => 1,
                "w" => 7,
                "mo" => 30,
                "y" => 365,
                _ => return None,
            };
            let days = u32::try_from(n.checked_mul(scale)?).ok()?;
            Some(QueryTerm::TouchedAge { negated, op, days })
        }
        _ => None,
    }
}

const FIELDS: [&str; 9] = [
    "lang",
    "owner",
    "in",
    "is",
    "has",
    "touched",
    "size",
    "completion",
    "collection",
];

#[must_use]
pub fn parse_query(input: &str) -> QueryAst {
    let mut terms = Vec::new();
    let mut ignored = Vec::new();

    for token in tokenize(input) {
        let negated = token.starts_with('-');
        let body: String = if negated {
            token.chars().skip(1).collect()
        } else {
            token.clone()
        };
        if body.is_empty() {
            continue;
        }

        // `split_once` rather than an index, so no byte offset is ever used to slice a string
        // that may hold multi-byte characters. A leading colon is not a field.
        let split = body
            .split_once(':')
            .filter(|(prefix, _)| !prefix.is_empty());
        let Some((prefix, raw_value)) = split else {
            let (value, _) = unquote(&body);
            terms.push(QueryTerm::Bare {
                negated,
                text: value.to_lowercase(),
            });
            continue;
        };

        let field = prefix.to_lowercase();
        if !FIELDS.contains(&field.as_str()) {
            if is_field_shaped(prefix) && !raw_value.starts_with("//") {
                ignored.push(IgnoredTerm {
                    text: token,
                    reason: IgnoredReason::UnknownField,
                });
            } else {
                let (value, _) = unquote(&body);
                terms.push(QueryTerm::Bare {
                    negated,
                    text: value.to_lowercase(),
                });
            }
            continue;
        }

        if field == "completion" {
            ignored.push(IgnoredTerm {
                text: token,
                reason: IgnoredReason::NotComputed,
            });
            continue;
        }

        let (value, quoted) = unquote(raw_value);
        match parse_field_term(&field, &value, quoted, negated) {
            Some(term) => terms.push(term),
            None => ignored.push(IgnoredTerm {
                text: token,
                reason: IgnoredReason::MalformedValue,
            }),
        }
    }

    QueryAst {
        grammar_version: QUERY_GRAMMAR_VERSION,
        terms,
        ignored,
    }
}

File: app/src/shared/query/parse.ts
import type { IgnoredTerm, QueryAst, QueryTerm, TextField } from './ast.js';
import {
  NEVER_EVALUATED,
  QUERY_GRAMMAR_VERSION,
  SIZE_UNIT_BYTES,
  TOUCHED_UNIT_DAYS,
  isHasAttribute,
  isIsFlag,
  isQueryField,
} from './grammar.js';
import { splitFieldValue, tokenizeQuery, unquote } from './tokenize.js';

const TEXT_FIELDS: readonly string[] = ['lang', 'owner', 'in', 'collection'];
const COMPARISON = /^([<>])(\d+)([a-z]+)$/;
const YEAR = /^\d{4}$/;
/** A prefix that reads as a field name: two or more letters. One letter is a drive letter. */
const FIELD_SHAPED = /^[A-Za-z]{2,}$/;

export function parseQuery(input: string): QueryAst {
  const terms: QueryTerm[] = [];
  const ignored: IgnoredTerm[] = [];

  for (const token of tokenizeQuery(input)) {
    const negated = token.text.startsWith('-');
    const body = negated ? token.text.slice(1) : token.text;
    if (body.length === 0) continue;

    const split = splitFieldValue(body);
    if (split === null) {
      terms.push({ kind: 'bare', negated, text: unquote(body).value.toLowerCase() });
      continue;
    }

    const field = split.field.toLowerCase();
    if (!isQueryField(field)) {
      if (FIELD_SHAPED.test(split.field) && !split.value.startsWith('//')) {
        ignored.push({ text: token.text, reason: 'unknownField' });
      } else {
        terms.push({ kind: 'bare', negated, text: unquote(body).value.toLowerCase() });
      }
      continue;
    }

    if (NEVER_EVALUATED.includes(field)) {
      ignored.push({ text: token.text, reason: 'notComputed' });
      continue;
    }

    const term = parseFieldTerm(field, split.value, split.quoted, negated);
    if (term === null) ignored.push({ text: token.text, reason: 'malformedValue' });
    else terms.push(term);
  }

  return { grammarVersion: QUERY_GRAMMAR_VERSION, terms, ignored };
}

function parseFieldTerm(
  field: string,
  raw: string,
  quoted: boolean,
  negated: boolean,
): QueryTerm | null {
  if (raw.length === 0) return null;

  if (TEXT_FIELDS.includes(field)) {
    const value = quoted ? raw : raw.toLowerCase();
    return { kind: 'text', negated, field: field as TextField, value, quoted };
  }

  if (field === 'is') {
    const flag = raw.toLowerCase();
    return isIsFlag(flag) ? { kind: 'flag', negated, flag } : null;
  }

  if (field === 'has') {
    const attribute = raw.toLowerCase();
    return isHasAttribute(attribute) ? { kind: 'has', negated, attribute } : null;
  }

  if (field === 'size') {
    const m = COMPARISON.exec(raw.toLowerCase());
    if (m === null) return null;
    const [, sign, digits, unit] = m;
    const scale = (SIZE_UNIT_BYTES as Record<string, number | undefined>)[unit ?? ''];
    if (scale === undefined || sign === undefined || digits === undefined) return null;
    return { kind: 'size', negated, op: sign === '>' ? 'gt' : 'lt', bytes: Number(digits) * scale };
  }

  // [p3] §31.1: a bare integer bound with no unit, because the quantity is a COUNT OF CHECKS —
  // `completion:>5kb` is not a thing to admit, so this parses its own two characters rather
  // than borrowing `COMPARISON`, which requires a unit.
  if (field === 'completion') {
    const m = /^([<>])(\d+)$/u.exec(raw.toLowerCase());
    const sign = m?.[1];
    const digits = m?.[2];
    if (sign === undefined || digits === undefined) return null;
    return { kind: 'completion', negated, op: sign === '>' ? 'gt' : 'lt', value: Number(digits) };
  }

  if (field === 'touched') {
    const lowered = raw.toLowerCase();
    if (YEAR.test(lowered)) return { kind: 'touchedYear', negated, year: Number(lowered) };
    const m = COMPARISON.exec(lowered);
    if (m === null) return null;
    const [, sign, digits, unit] = m;
    const scale = (TOUCHED_UNIT_DAYS as Record<string, number | undefined>)[unit ?? ''];
    if (scale === undefined || sign === undefined || digits === undefined) return null;
    return {
      kind: 'touchedAge',
      negated,
      op: sign === '>' ? 'gt' : 'lt',
      days: Number(digits) * scale,
    };
  }

  return null;
}
